domain/resource/state/resource.go

```diff
@@ -407,173 +419,568 @@ WHERE uuid = $resourceIdentity.uuid`,
-FROM resource_retrieved_by
-WHERE resource_uuid = $resourceIdentity.uuid`
-	checkAlreadyRetrievedStmt, err := st.Prepare(checkAlreadyRetrievedQuery, resourceUUID)
+FROM   resource_retrieved_by
+WHERE  resource_uuid = $resourceIdentity.uuid`
+	checkAlreadyRetrievedStmt, err := st.Prepare(checkAlreadyRetrievedQuery, resID)
 	if err != nil {
-		return resource.SetUnitResourceResult{}, errors.Capture(err)
+		return errors.Capture(err)
 	}
 
-	// Prepare statements to update retrieved data if not already retrieved.
-	type retrievedByType struct {
+	err = tx.Query(ctx, checkAlreadyRetrievedStmt, resID).Get(&resID)
+	if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+		return errors.Capture(err)
+	} else if err == nil {
+		// If the resource has already been retrieved, the return an error.
+		return resourceerrors.ResourceAlreadyStored
+	}
+
+	// Get the retrieved by type ID.
+	type getRetrievedByType struct {
 		ID   int    `db:"id"`
 		Name string `db:"name"`
 	}
-	type retrievedBy struct {
+	retrievedTypeParam := getRetrievedByType{Name: string(retrievedByType)}
+	getRetrievedByTypeStmt, err := st.Prepare(`
+SELECT &getRetrievedByType.* 
+FROM   resource_retrieved_by_type 
+WHERE  name = $getRetrievedByType.name`, retrievedTypeParam)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	err = tx.Query(ctx, getRetrievedByTypeStmt, retrievedTypeParam).Get(&retrievedTypeParam)
+	if errors.Is(err, sqlair.ErrNoRows) {
+		return resourceerrors.RetrievedByTypeNotValid
+	}
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	// Insert retrieved by.
+	type setRetrievedBy struct {
 		ResourceUUID      string `db:"resource_uuid"`
 		RetrievedByTypeID int    `db:"retrieved_by_type_id"`
 		Name              string `db:"name"`
 	}
-	retrievedTypeParam := retrievedByType{Name: string(config.RetrievedByType)}
-	retrievedByParam := retrievedBy{ResourceUUID: config.ResourceUUID.String(), Name: config.RetrievedBy}
-	getRetrievedTypeQuery := `
-	SELECT &retrievedByType.* 
-	FROM resource_retrieved_by_type 
-	WHERE name = $retrievedByType.name`
-	getRetrievedByTypeStmt, err := st.Prepare(getRetrievedTypeQuery, retrievedTypeParam)
-	if err != nil {
-		return resource.SetUnitResourceResult{}, errors.Capture(err)
+	retrievedByParam := setRetrievedBy{
+		ResourceUUID:      resourceUUID.String(),
+		Name:              retrievedBy,
+		RetrievedByTypeID: retrievedTypeParam.ID,
 	}
-	insertRetrievedByQuery := `
+	insertRetrievedByStmt, err := st.Prepare(`
 INSERT INTO resource_retrieved_by (resource_uuid, retrieved_by_type_id, name)
-VALUES ($retrievedBy.*)`
-	insertRetrievedByStmt, err := st.Prepare(insertRetrievedByQuery, retrievedByParam)
+VALUES      ($setRetrievedBy.*)`, retrievedByParam)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	return errors.Capture(tx.Query(ctx, insertRetrievedByStmt, retrievedByParam).Run())
+}
+
+// incrementCharmModifiedVersion increments the charm modified version on any
+// application associated with a resource.
```

Comment:
**thought:** feels strange, because in reality this resource are associated with 1 app, and only 1. Even if it is accurate that it may be the case, regarding the DDL